src/Rings/AbelianClosure.jl

```diff
@@ -0,0 +1,777 @@
+# even and n%4!=0. either we can construct the field Q(z_n) or we take -z_(n/2)
+# as a primitive n-th root. to change between these two options, use
+# PCharSaturateAll with allroots or allrootsNew (change this in the code)
+
+module AbelianClosure 
+
+using ..Oscar
+
+import Base: +, *, -, //, ==, zero, one, ^, div, isone, iszero, deepcopy_internal
+
+#import ..Oscar.AbstractAlgebra: promote_rule
+
+import ..Oscar: addeq!, isunit, parent_type, elem_type, gen, root_of_unity, root, divexact, mul!, roots, isroot_of_unity, promote_rule, AbstractAlgebra
+
+export abelian_closure, QabField, QabElem, set_variable!, get_variable, make_compatible
+
+################################################################################
+#
+#  Types
+#
+################################################################################
+
+mutable struct QabField <: Nemo.Field # union of cyclotomic fields
+  s::String
+  fields::Dict{Int, AnticNumberField} # Cache for the cyclotomic fields
+end
+
+const _Qab = QabField("z", Dict{Int, AnticNumberField}())
+
+mutable struct QabElem <: Nemo.FieldElem
+  data::nf_elem                       # Element in cyclotomic field
+  c::Int                              # Conductor of field
+end
+
+# This is a functor like object G with G(n) = primitive n-th root of unity
+
+mutable struct QabFieldGen
+  K::QabField
+end
+
+const _QabGen = QabFieldGen(_Qab)
+
+################################################################################
+#
+#  Creation of the field
+#
+################################################################################
+
+"""
+    abelian_closure(QQ::RationalField)
+
+Return a pair `(K, z)` consisting of the abelian closure `K` of the rationals
+and a generator `z` that can be used to construct primitive roots of unity in
+`K`.
+"""
+abelian_closure(::FlintRationalField) = _Qab, _QabGen
+
+"""
+    gen(K::QabField)
+
+Return the generator of the abelian closure `K` that can be used to construct
+primitive roots of unity.
+"""
+gen(K::QabField) = _QabGen
+
+"""
+    gen(K::QabField, s::String)
+
+Return the generator of the abelian closure `K` that can be used to construct
+primitive roots of unity. The string `s` will be used during printing.
+"""
+function gen(K::QabField, s::String)
+  K.s = s
+  return gen(K)
+end
+
+################################################################################
+#
+#  Parent and element functions
+#
+################################################################################
+
+elem_type(::Type{QabField}) = QabElem
+
+elem_type(::QabField) = QabElem
+
+parent_type(::Type{QabElem}) = QabField
+
+parent_type(::QabElem) = QabField
+
+Oscar.parent(::QabElem) = _Qab
+
+################################################################################
+#
+#  Field access
+#
+################################################################################
+
+_variable(K::QabField) = K.s
+
+_variable(b::QabElem) = string(_variable(_Qab), "(", b.c, ")")
```

Comment:
For expressification purposes, it would be best if this was
`Expr(:call, Symbol(_variable(_Qab)), b.c)`

Reply:
Will do. Will also use the name macro.